pep-9999.rst

```diff
@@ -0,0 +1,180 @@
+    Found 1 error in 1 file (checked 1 source file)
+
+
+Specification
+=============
+
+When a module object is called, and a ``__call__`` function is found (either
+as the result of a ``__getattr__`` or ``__dict__`` lookup), then that function
+will be called with the given arguments.
+
+If a ``__call__`` object is not found, then a ``TypeError`` will be raised,
+matching the existing behavior.
+
+All of these examples would be considered valid, callable modules:
+
+.. code-block:: python
+
+    # hello.py
+
+    def __call__(...):
+        pass
+
+.. code-block:: python
+
+    # hello.py
+
+    class Hello:
+        pass
+
+    __call__ = Hello
+
+.. code-block:: python
+
+    # hello.py
+
+    def hello():
+        pass
+
+    def __getattr__(name):
+        if name == "__call__":
+            return hello
+
+The first two styles should generally be preferred, as it allows for easier
+static analysis from tools like type checkers, though the third form would be
+allowed in order to make the implementation more consistent.
+
+
+Backwards compatibility and impact on performance
+=================================================
+
+This PEP is not expected to cause any backwards incompatibility. Any modules
+that already contain a ``__call__`` member will continue to function the same
+as before, though with the additional ability to be called directly. It is
+considered unlikely that modules with an existing ``__call__`` member would
+depend on the existing behavior of raising ``TypeError`` when called.
+
+Performance implications of this PEP are minimal, as it defines a new interface.
+Calling a module would trigger a lookup for the name ``__call__`` on a module
+object. Existing workarounds for creating callable modules already depend on
+this behavior for generic objects, resulting in similar performance for these
+callable modules.
+
+Type checkers will likely need to be updated accordingly to treat modules with
+a ``__call__`` function as callable. This should be possible to support in type
+checkers when checking code targeted at older Python versions that do not
+support callable modules, with the expectation that these modules would also
+include one of the workarounds mentioned earlier to make the module callable.
+
+
+Rejected Ideas
+==============
+
+Given the introduction of ``__getattr__`` and ``__dir__``, and the proposal
+to enable use of ``__call__``, it was considered if it was worth allowing use
+of *all* special methods [1]_ for modules, such as ``__or__`` and ``__iter__``.
+While this would not be completely undesired, it increases the potential for
+backward compatibility concerns, and these other special methods are likely
+to provide less utility to library authors in comparison to ``__call__``.
+
+
+References
+==========
+
+.. [1] Data model, special methods
+   (https://docs.python.org/3/reference/datamodel.html#special-method-names)
+
+
+Copyright
+=========
+
+This document has been placed in the public domain.
```

Comment:
Sorry to add another comment thread! Just spotted that this is using the old copyright template —- please update to the new CC0 one. Likely as a result, the teaching and security [suggested sections](https://peps.python.org/pep-0012/#suggested-sections) are missing – though if you don’t feel they’re needed, they aren’t mandatory.

A